packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -0,0 +1,3450 @@
+      typedef Tpetra::RowMatrix<scalar_type,local_ordinal_type,global_ordinal_type,node_type> tpetra_row_matrix_type;
+      typedef Tpetra::BlockCrsMatrix<scalar_type,local_ordinal_type,global_ordinal_type,node_type> tpetra_block_crs_matrix_type;
+      typedef typename tpetra_block_crs_matrix_type::little_block_type tpetra_block_access_view_type;
+      typedef Tpetra::BlockMultiVector<scalar_type,local_ordinal_type,global_ordinal_type,node_type> tpetra_block_multivector_type;
+      typedef typename tpetra_block_crs_matrix_type::crs_graph_type::local_graph_type local_crs_graph_type;
+
+      ///
+      /// simd vectorization
+      ///
+      template<typename T, int l> using Vector = KokkosBatched::Experimental::Vector<T,l>;
+      template<typename T> using SIMD = KokkosBatched::Experimental::SIMD<T>;
+      template<typename T, typename M> using DefaultVectorLength = KokkosBatched::Experimental::DefaultVectorLength<T,M>;
+
+      // Kyungjoo: hansen enum does not work (don't know why)
+      // enum : int { vector_length = DefaultVectorLength<impl_scalar_type,memory_space>::value };
+      static constexpr int vector_length = DefaultVectorLength<impl_scalar_type,memory_space>::value;
+      typedef Vector<SIMD<impl_scalar_type>,vector_length> vector_type;
+
+      ///
+      /// commonly used view types 
+      ///
+      typedef Kokkos::View<size_type*,device_type> size_type_1d_view;
+      typedef Kokkos::View<local_ordinal_type*,device_type> local_ordinal_type_1d_view;
+      // tpetra block crs values
+      typedef Kokkos::View<impl_scalar_type*,device_type> impl_scalar_type_1d_view;
+      // tpetra multivector values (layout left): may need to change the typename more explicitly 
+      typedef Kokkos::View<impl_scalar_type**,Kokkos::LayoutLeft,device_type> impl_scalar_type_2d_view;
+      // packed data always use layout right
+      typedef Kokkos::View<vector_type*,device_type> vector_type_1d_view;
+      typedef Kokkos::View<vector_type***,Kokkos::LayoutRight,device_type> vector_type_3d_view;
+      typedef Kokkos::View<impl_scalar_type****,Kokkos::LayoutRight,device_type> impl_scalar_type_4d_view;
+    };
+    
+    ///
+    /// setup sequential importer
+    ///
+    template<typename MatrixType>
+    typename Teuchos::RCP<const typename ImplType<MatrixType>::tpetra_import_type> 
+    createBlockCrsTpetraImporter(const Teuchos::RCP<const typename ImplType<MatrixType>::tpetra_block_crs_matrix_type> &A) {
+      using impl_type = ImplType<MatrixType>;
+      using tpetra_map_type = typename impl_type::tpetra_map_type;
+      using tpetra_mv_type = typename impl_type::tpetra_block_multivector_type;
+      using tpetra_import_type = typename impl_type::tpetra_import_type;
+
+      const auto g = A->getCrsGraph();  // tpetra crs graph object
+      const auto blocksize = A->getBlockSize();
+      const auto src = Teuchos::rcp(new tpetra_map_type(tpetra_mv_type::makePointMap(*g.getDomainMap(), blocksize)));
```

Comment:
I remember working with @ambrad on this issue.  Communication _should_ be able to use the mesh Maps, but for some reason BlockMultiVector communication was slow and it was better to use the point Maps.  Is that still the case?

Reply:
In this PR, the primary goal is to port the ifpack container to GPUs. Regarding to communication, I do not test but use the same pattern as previous implementation.